File: lab6/c/airports.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdarg.h>
#include <stdbool.h>
#include <assert.h>
#include <math.h>

#include "hashmap.h"
#include "general.h"
#include "graph.h"

#include "airports.h"


#define R 6371
#define TO_RAD (3.1415926536 / 180)
double dist(double lat1, double lng1, double lat2, double lng2)
{
	double dx, dy, dz;
	lng1 -= lng2;
	lng1 *= TO_RAD, lat1 *= TO_RAD, lat2 *= TO_RAD;

	dz = sin(lat1) - sin(lat2);
	dx = cos(lng1) * cos(lat1) - cos(lat2);
	dy = sin(lng1) * cos(lat1);
	return asin(sqrt(dx * dx + dy * dy + dz * dz) / 2) * 2 * R;
}

typedef struct {
  char code[4]; // We use the empty string to indicate unassigned ids
  double lat;
  double lng;
} airport_t;

size_t num_apids = 0;
airport_t *airports = NULL;
hashmap_t *code_id_map = NULL;

graph_t *route_g = NULL;


size_t ap_get_num_ids() {return num_apids;}


void ap_read(FILE *f) {
  size_t n;

  assert(airports == NULL && code_id_map==NULL);

  if (fscanf(f,"%zu",&num_apids)!=1) error("Error reading airport file");
  if (fscanf(f,"%zu",&n)!=1) error("Error reading airport file");

  airports = CALLOC(airport_t, num_apids);

  code_id_map = hashmap_new(num_apids,HASH_1_LINEAR_PROBING);

  for (size_t i=0;i<n;++i) {

    apid_t id;
    char code[4];
    double lat;
    double lng;

    if (fscanf(f,"%zu %3s %lf %lf\n",&id,(char*)&code,&lat,&lng) != 4)
        error("Error reading airport file");

    if (!(id<num_apids)) error("Invalid airport ID: %zu", id);

    airports[id] = (airport_t){{code[0],code[1],code[2],0},lat,lng};


    hashmap_insert(code,id,code_id_map);
  }

//   printf("Read %zu airports\n",n);

}

void ap_read_routes(FILE *f) {
  size_t num_edges = scan_size(f);

  assert(route_g == NULL);
  route_g = graph_new(ap_get_num_ids());

  for (int i=0;i<num_edges;++i) {
    apid_t id1 = ap_check_id(scan_size(f));
    apid_t id2 = ap_check_id(scan_size(f));

    weight_t dist = ap_get_dist(id1,id2);

    graph_add_edge(route_g,id1,dist,id2);
  }
}

FILE *fopenchk(char const *name, char const *mode) {
  FILE *f = fopen(name,mode);
  if (!f) error("Error opening '%s' for %s",name,mode);
  return f;
}

void ap_std_init() {
  {
    FILE *f = fopenchk("../data/airports.txt","r");
    ap_read(f);
    fclose(f);
  }

  {
    FILE *f = fopenchk("../data/routes.txt","r");
    ap_read_routes(f);
    fclose(f);
  }
}



graph_t const *ap_get_graph() {return route_g;}

weight_t ap_get_dist(apid_t id1, apid_t id2) {
  return weight_of_int(round(ap_get_dist_dbl(id1,id2)));
}



void ap_delete() {
  if (airports) free(airports);
  if (code_id_map) hashmap_delete(code_id_map);
  if (route_g) graph_delete(route_g);
}

bool ap_is_valid_id(apid_t id) {
   // printf("id: %zu",id);
  //  printf("code: %s", airports[id].code);
  return id<num_apids && airports[id].code[0]!=0;
}

void ap_invalidate_id(apid_t id) {
  assert(id < num_apids);
  // Note: As our hashtables do not support delete, we insert a validity check after retrieving a code.
  airports[id].code[0] = 0;
}


apid_t ap_check_id(apid_t id) {
  if (!ap_is_valid_id(id)) error("Invalid airport id: %zu",id);
  return id;
}


char *ap_get_code(apid_t id) {
  assert(ap_is_valid_id(id));
  return airports[id].code;
};
double ap_get_lat(apid_t id) {
  assert(ap_is_valid_id(id));
  return airports[id].lat;
};

double ap_get_lng(apid_t id) {
  assert(ap_is_valid_id(id));
  return airports[id].lng;
};

bool ap_get_id_aux(char const *code, apid_t *res) {
  return (hashmap_lookup(code,res,code_id_map) && ap_is_valid_id(*res));
}

bool ap_is_valid_code(char const *code) {
  apid_t res;
  return ap_get_id_aux(code,&res);
}

apid_t ap_get_id(char const *code) {
  apid_t res;
  if (!ap_get_id_aux(code,&res)) error("Unknown airport ID: %s",code);

  return res;
}

double ap_get_dist_dbl(apid_t id1, apid_t id2) {
  return dist(ap_get_lat(id1), ap_get_lng(id1), ap_get_lat(id2), ap_get_lng(id2));
}

File: lab6/python/airports.py
import math

from hashmap import hashmap_t, HashingModes
import general
from graph import graph_t
import weight

R = 6371
TO_RAD = (3.1415926536 / 180)

def dist(lat1, lng1, lat2, lng2):
    lng1 = lng1 - lng2
    lng1 = lng1 * TO_RAD
    lat1 = lat1 * TO_RAD
    lat2 = lat2 * TO_RAD
    
    dz = math.sin(lat1) - math.sin(lat2)
    dx = math.cos(lng1) * math.cos(lat1) - math.cos(lat2)
    dy = math.sin(lng1) * math.cos(lat1)
    return math.asin(math.sqrt(dx * dx + dy * dy + dz * dz) / 2) * 2 * R
    
class airport_t:
    def __init__(self, code, lat, lng):
        self.code = code
        self.lat = lat
        self.lng = lng
        
num_apids = 0
airports = None
code_id_map = None

route_g = None

##
 # Get maximum id + 1
 #/
def ap_get_num_ids():
    return num_apids
    
## Read airport list from file.
 # The file has the following format:
 #
 # num_ids       // Maximum Id + 1
 # num_airports  // Number of actual airport entries in file
 # Id IATA Lat Long // Entry for an airport.
 # ...
 #
 #
def ap_read(f):
    global airports, code_id_map, num_apids
    
    assert (airports == None and code_id_map == None)
    
    num_apids = int(f.readline())
    n = int(f.readline())
    
    airports = [None]*num_apids
    
    code_id_map = hashmap_t(num_apids, HashingModes.HASH_1_LINEAR_PROBING)
    
    for i in range(0, n):
        (id_string, code, lat_string, lng_string) = f.readline().split()
        
        id = int(id_string)
        
        if (not id < num_apids):
            general.error("Invalid airport ID: %d", id)
            
        airports[id] = airport_t(code, float(lat_string), float(lng_string))
        
        code_id_map.hashmap_insert(code, id)
        
##
 # Read routes from file.
 #/
def ap_read_routes(f):
    global route_g
    
    num_edges = int(f.readline())
    
    assert (route_g == None)
    
    route_g = graph_t(ap_get_num_ids())
    
    for i in range(0, num_edges):
        (id1_string, id2_string) = f.readline().split()
        id1 = ap_check_id(int(id1_string))
        id2 = ap_check_id(int(id2_string))
        
        dist = ap_get_dist(id1, id2)
        
        route_g.graph_add_edge(id1, dist, id2)
        
def fopenchk(name, mode):
    try:
        f = open(name, mode)
        return f
    except:
        general.error("Error opeining %s for %s" % (name, mode))
        
##
 # Reads airport list and routes from files "airports.txt" and "routes.txt", which must be in the current directory.
 #/
def ap_std_init():
    f = fopenchk("../data/airports.txt", "r")
    ap_read(f)
    f.close()
    
    f = fopenchk("../data/routes.txt", "r")
    ap_read_routes(f)
    f.close()

##
 # Returns a graph where the nodes are airport IDs, and the edges represent connections. They are labeled with the distance in km.
 #
 # @note The graph is constructed by @ref ap_read_routes, this function only returns it, and is thus O(1).
 #
 #/
def ap_get_graph():
    return route_g
    
##
 # Get the distance between two airports as weight.
 #/
def ap_get_dist(id1, id2):
    return weight.weight_t(round(ap_get_dist_dbl(id1, id2)))

##
 # Return true if id is actually assigned to an airport.
 # That is, id < num_ids, and there is an airport entry for this id.
 #/
def ap_is_valid_id(id):
    return id < num_apids and airports[id] != None and airports[id].code != "0"

##
 # Remove the entry for an airport.
 # @pre id < num_ids, but not necessarily valid!
 #/
def ap_invalidate_id(id):
    assert(id < num_apids)
    # Note: As our hashtables do not support delete, we insert a validity check after retrieving a code
    airports[id].code = "0"

##
 # Exit on invalid id, or return unchanged id.
 # @pre id must be valid
 #/
def ap_check_id(id):
    if (not ap_is_valid_id(id)):
        general.error("Invalid aiport id: %d" % id)
    return id

# id must be valid
def ap_get_code(id):
    assert (ap_is_valid_id(id))
    return airports[id].code
    
# id must be valid
def ap_get_lat(id):
    assert (ap_is_valid_id(id))
    return airports[id].lat
    
# id must be valid
def ap_get_lng(id):
    assert (ap_is_valid_id(id))
    return airports[id].lng
    
def ap_get_id_aux(code):
    return code_id_map.hashmap_lookup(code)
    
def ap_is_valid_code(code):
    res = None
    res = ap_get_id_aux(code)
    return res != None and ap_is_valid_id(res)

/##
 # Exits with defined error message on invalid code.
 #/
def ap_get_id(code):
    res = ap_get_id_aux(code)
    if (res == None or not (ap_is_valid_id(res))):
        general.error("Unknown airport ID: %s" % code)
    return res

##
 # Get the distance, in km, between two airports.
 # This uses the haversine-formula, regarding the earth as sphere with radius 6371km.
 # @pre ids must be valid
 #/
def ap_get_dist_dbl(id1, id2):
    return dist(ap_get_lat(id1), ap_get_lng(id1), ap_get_lat(id2), ap_get_lng(id2))
